File: src/com/lysdev/transperthcached/activities/train/TrainStationTimesActivity.java
package com.lysdev.transperthcached.activities.train;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.widget.ListView;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;

import java.util.ArrayList;
import java.util.List;

import org.joda.time.Minutes;
import org.joda.time.DateTime;

import com.lysdev.transperthcached.livetimes.Trip;
import com.lysdev.transperthcached.livetimes.GetTimesForStation;
import com.lysdev.transperthcached.livetimes.TimesForStation;
import com.lysdev.transperthcached.livetimes.InvalidPlatformCodeException;
import com.lysdev.transperthcached.R;


class TripDisplayWrapper {
    Trip t;
    public TripDisplayWrapper(Trip t) {
        this.t = t;
    }
    public String toString() {
        return String.format(
            "%s - %d minutes - %d cars",
            this.t.getLineFull(),
            Minutes.minutesBetween(
                DateTime.now(),
                this.t.getActual()
            ).getMinutes(),
            this.t.getNumCars()
        );
    }
}


public class TrainStationTimesActivity extends Activity {
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.station_times);

        String line_name = getIntent().getStringExtra("line_name");
        String station_name = getIntent().getStringExtra("station_name");
        Direction direction = Direction.from_val("direction", getIntent());

        display_data(line_name, station_name, direction);
    }

    public String b_to_s(boolean b) {
        return b ? "true" : "false";
    }

    public void display_data(String line_name, String station_name, Direction direction) {
        TimesForStation tfp = GetTimesForStation.getTimes(station_name);

        List<Trip> trips = tfp.getTrips();
        List<TripDisplayWrapper> filtered = new ArrayList<TripDisplayWrapper>();

        for (Trip t : trips) {
            boolean going, coming;

            Log.d(
                "TransperthCached",
                String.format(
                    "%s %s",
                    t.getDestination(),
                    direction.toString()
                )
            );

            coming = t.getDestination().equals("Perth") && direction == Direction.TO;
            going =  !t.getDestination().equals("Perth") && direction == Direction.FROM;

            Log.d("TransperthCached", String.format("going: %s", b_to_s(going)));
            Log.d("TransperthCached", String.format("coming: %s", b_to_s(coming)));

            if (going || coming) filtered.add(new TripDisplayWrapper(t));
        }

        Log.d(
            "TransperthCached",
            String.format(
                "%d before, %d now",
                trips.size(),
                filtered.size()
            )
        );

        ListView times_lv = (ListView) findViewById(R.id.times);
        ListAdapter ad = new ArrayAdapter<TripDisplayWrapper>(
            this,
            android.R.layout.simple_list_item_1,
            android.R.id.text1,
            filtered
        );
        times_lv.setAdapter(ad);
        // times_lv.setOnItemClickListener(this);
    }
}

File: src/com/lysdev/transperthcached/activities/train/TrainStationTimesActivity.scala
package com.lysdev.transperthcached.activities.train

import java.util.ArrayList

import android.app.ProgressDialog
import android.os.Bundle
import android.os.AsyncTask
import android.util.Log
import android.os.Looper
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.{AdapterView, ArrayAdapter, ListAdapter, TextView, ListView, Toast}

import org.scaloid.common._
import scala.collection.JavaConverters._
import scala.collection.JavaConversions._

import org.joda.time.{Minutes, DateTime}

import com.lysdev.transperthcached.livetimes.{
    Trip,
    GetTimesForStation,
    TimesForStation,
    InvalidPlatformCodeException
}
import com.lysdev.transperthcached.R


case class TripDisplayWrapper(trip: Trip) {
    override def toString() : String = {
        var s : String = this.trip.getLineFull()
        val mins : Integer = Minutes.minutesBetween(
            DateTime.now(),
            this.trip.getActual()
        ).getMinutes
        val cars : Integer = this.trip.getNumCars

        String.format(
            "%s - %d minutes - %d cars",
            s, mins, cars
        )
    }
}


class FetchTimesTask[V <: android.view.View, T <: AnyRef]
                    (var klass : TrainStationTimesActivity,
                     var direction : Direction,
                     var ad : ArrayAdapter[TripDisplayWrapper])
                     extends AsyncTask[AnyRef, Void, List[Trip]] {
    var mDialog : ProgressDialog = null

    override protected def onPreExecute() {
        this.mDialog = new ProgressDialog(this.klass)
        this.mDialog.setMessage("Loading...")
        this.mDialog.setCancelable(true)
        this.mDialog.show()
    }

    override protected def doInBackground(p1: AnyRef*) : List[Trip] = {
        (
            GetTimesForStation
            .getTimes(p1.head.asInstanceOf[String])
            .getTrips().toList
        )
    }

    override protected def onPostExecute(trips: List[Trip]) {
        Log.d(
            "TransperthCached",
            String.format(
                "%s trips",
                trips.length.toString()
            )
        )

        this.klass.filtered = (
            trips
            .filter(trip => {
                val to_perth = trip.getDestination().equals("Perth")
                val coming = to_perth && this.direction == Direction.TO
                val going = !to_perth && this.direction == Direction.FROM

                going || coming
            })
            .map(new TripDisplayWrapper(_))
        )

        this.ad.clear()
        this.ad.addAll(this.klass.filtered.asJavaCollection)
        this.ad.notifyDataSetChanged()

        this.mDialog.dismiss()
        Log.d("TransperthCached", "initialized")
    }
}


class TrainStationTimesActivity extends SActivity
                                        with OnItemClickListener {
    var filtered : List[TripDisplayWrapper] = null
    lazy val times_lv = find[ListView](R.id.times)

    override def onCreate(savedInstanceState: Bundle) = {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.station_times)

        val line_name = getIntent().getStringExtra("line_name")
        val station_name = getIntent().getStringExtra("station_name")
        val direction = Direction.from_val("direction", getIntent())
        this.filtered = List[TripDisplayWrapper]()

        display_data(line_name, station_name, direction)
    }

    def display_data(line_name: String, station_name: String, direction: Direction) = {
        // https://stackoverflow.com/questions/3200551/unable-to-modify-arrayadapter-in-listview-unsupportedoperationexception
        val ad = new ArrayAdapter(
            this,
            android.R.layout.simple_list_item_1,
            new ArrayList(this.filtered.asJavaCollection)
        )
        this.times_lv.setOnItemClickListener(this)
        this.times_lv.setAdapter(ad)

        (
            new FetchTimesTask[TextView, TripDisplayWrapper](this, direction, ad)
            .execute(station_name)
        )
    }

    def onItemClick(parent: AdapterView[_], view: View, position: Int, id: Long) : Unit = {
        val trp = this.filtered(position).trip

        Log.d("TransperthCached", String.format("Clicked: %s", new TripDisplayWrapper(trp)))

        Toast.makeText(
            this,
            trp.getPatternFullDisplay().asScala.mkString(", "),
            Toast.LENGTH_LONG
        ).show()
    }
}
